File: Binary Search/Hard/LET_no644_Maximum Average Subarray II.cpp
class Solution {
public:
    double findMaxAverage(vector<int>& nums, int k) {
        double min_val = *min_element(nums.begin(), nums.end()) * 1.0;
        double max_val = *max_element(nums.begin(), nums.end()) * 1.0;

        while (min_val + 1e-5 <= max_val) {
            double mid_val = min_val + (max_val - min_val) / 2;
            if (hasSubarray(nums, k, mid_val))
                min_val = mid_val;
            else
                max_val = mid_val;
        }

        if (hasSubarray(nums, k, max_val))
            return max_val;
        return min_val;
    }

    bool hasSubarray(vector<int>& nums, int k, double average) {
        double cur = 0.0, pre = 0.0, pre_min = 0.0;
        for (int i = 0; i < k - 1; i++)
            cur += nums[i] - average;
        
        for (int i = k - 1; i < nums.size(); i++) {
            cur += nums[i] - average;
            if (cur >= pre_min)
                return true;
            
            pre += nums[i - k + 1] - average;
            pre_min = min(pre_min, pre);
        }

        return false;
    }
};

File: Binary Search/Medium/LET_no644_Maximum Average Subarray II.py
class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        min_val, max_val = min(nums), max(nums)
        while min_val + 1e-5 <= max_val:
            mid = (min_val + max_val) / 2
            if self.is_valid(nums, mid, k):
                min_val = mid
            else:
                max_val = mid

        return max_val if self.is_valid(nums, max_val, k) else min_val

    def is_valid(self, nums, avg, k):
        cur = sum(nums[:k - 1]) - avg * (k - 1)
        pre = pre_min = 0
        
        for i in range(k - 1, len(nums)):
            cur += nums[i] - avg
            if cur - pre_min >= 0:
                return True
            
            pre += nums[i - k + 1] - avg
            pre_min = min(pre_min, pre)
        
        return False